core/src/main/java/com/linecorp/armeria/common/logging/LoggingDecoratorBuilder.java

```diff
@@ -555,104 +593,33 @@ protected final LogWriter logWriter() {
         if (logWriter != null) {
             return logWriter;
         }
-        if (!buildLogWriter) {
-            if (logger != null) {
-                return LogWriter.of(logger);
+
+        if (logWriterBuilder == null) {
+            // Neither logWriter nor log properties are set.
+            if (defaultLogger != null) {
+                return LogWriter.of(defaultLogger);
             } else {
                 return LogWriter.of();
             }
         }
-        final LogFormatter logFormatter =
-                LogFormatter.builderForText()
-                            .requestHeadersSanitizer(convertToStringSanitizer(requestHeadersSanitizer))
-                            .responseHeadersSanitizer(convertToStringSanitizer(responseHeadersSanitizer))
-                            .requestTrailersSanitizer(convertToStringSanitizer(requestTrailersSanitizer))
-                            .responseTrailersSanitizer(convertToStringSanitizer(responseTrailersSanitizer))
-                            .requestContentSanitizer(convertToStringSanitizer(requestContentSanitizer))
-                            .responseContentSanitizer(convertToStringSanitizer(responseContentSanitizer))
-                            .build();
-        final LogWriterBuilder builder = LogWriter.builder();
-        builder.logFormatter(logFormatter);
-        if (logger != null) {
-            builder.logger(logger);
-        }
 
-        if (requestLogLevelMapper != null) {
-            builder.requestLogLevelMapper(requestLogLevelMapper);
+        assert logFormatterBuilder != null;
+        final LogFormatter logFormatter = logFormatterBuilder.build();
+        logWriterBuilder.logFormatter(logFormatter);
+        if (logWriterBuilder.logger() == null && defaultLogger != null) {
+            logWriterBuilder.logger(defaultLogger);
         }
-        if (responseLogLevelMapper != null) {
-            builder.responseLogLevelMapper(responseLogLevelMapper);
-        }
-        builder.responseCauseFilter(responseCauseFilter);
-        return builder.build();
+        return logWriterBuilder.build();
     }
 
-    private void setBuildLogWriter() {
+    private void maybeCreateLogWriterBuilder() {
         if (logWriter != null) {
             throw new IllegalStateException("The logWriter and the log properties cannot be set together.");
         }
-        buildLogWriter = true;
-    }
-
-    @Override
-    public String toString() {
```

Comment:
Removed this since we reached a consensus that `toString()` is not a mandatory implementation for builder classes.